vsr: extra test for checksum

matklad opened the pull request:
vsr.checksum is pretty important, as it cements our on-disk representation and over-the-wire protocol (so, we won't be able to easily change it).

This PR adds a "change detector" test to ensure we don't accidentally change the result. The idea of the test is basically to hash a bunch of random stuff, and then assert that the final hash has a specific value.

While we are at it, extend our fuzz test to do more sanity checking:

- allow arbitrary bitflips (rather than just +%= 1)
- also test avalanching (that every bit of input affects every bit of the output)

While doing all that, I've noticed that the results we get are actually different from the results in Zig master. That's because we treat bytes-to-hash as a secret message, rather than as associated data. Adjust our algo to do that. I've also committed&reverted test against Zig master, which passes the change-detector test with updated expected value.

matklad commented:
Should be ready for review @kprotty! Dotnet tests fail, but that seems to be pre-existing problem on main which is being fixed independently.

kprotty commented on src/vsr/checksum.zig:
```diff
@@ -135,28 +138,85 @@ test "Aegis simple fuzzing" {
+
+    // Sanity check that we are not getting trivial answers.
+    for (cases) |case_a, i| {
+        assert(case_a != 0);
+        assert(case_a != std.math.maxInt(u128));
```
The zero case makes sense, but never considered if the full bits pattern was unlikely too.

matklad committed:
vsr: extra test for checksum

vsr.checksum is pretty important, as it cements our on-disk
representation and over-the-wire protocol (so, we won't be able to
easily change it).

This PR adds a "change detector" test to ensure we don't accidentally
change the result. The idea of the test is basically to hash a bunch of
random stuff, and then assert that the final hash has a specific value.

While we are at it, extend our fuzz test to do more sanity checking:

- allow arbitrary bitflips (rather than just +%= 1)
- also test avalanching (that every bit of input affects every bit of
  the output)

This actually uncovered  bug in the implementation: we were passing a
message as secret text to AEAD, but we should have passed it as
associated data, as the zig stdlib does.

Additionally, while this PR was in flight, we upgraded to Zig 0.11,
so switch to that implementation as well.

matklad commented:
Remixed this to use the upstream impl!

kprotty commented:
Looks good. Good enough that it needs to be off `tidy.zig`'s naughty list for the test/CI.

matklad commented:
Fixed!

matklad commented:
wait, no, test legitimately fail...

matklad commented:
Ok, now it seems to actually be green, ptal @kprotty

kprotty commented on src/vsr/checksum.zig:
```diff
@@ -1,132 +1,89 @@
+const assert = std.debug.assert;
 
-// TODO(king): Replace with std.crypto.aead.aegis.Aegis128LMac_128 once Zig is updated.
-const Aegis128 = struct {
```
Side note: wondering if we'll need to bring this back for checksum() to work at comptime (thinking `@inComptime()` switch for `aes.Block`)
